src/guard/config.ts

```diff
@@ -0,0 +1,291 @@
+import { DEFAULT_THRESHOLD, Threshold } from "./threshold.js";
+import { GuardName } from "./constants.js";
+import { GuardConfigValidationError, GuardConfigError } from "./errors.js";
+
+export class Match {
+    readonly threshold: Threshold;
+    readonly redact: boolean;
+    readonly count_threshold: number;
+
+    private constructor(
+        threshold: Threshold,
+        redact: boolean = false,
+        count_threshold: number = 0
+    ) {
+        this.threshold = threshold;
+        this.redact = redact;
+        this.count_threshold = count_threshold;
+        Object.freeze(this);
+    }
+
+    static create(
+        threshold: string | Threshold = DEFAULT_THRESHOLD,
+        redact: boolean = false,
+        count_threshold: number = 0
+    ): Match {
+        const thresholdObj = typeof threshold === 'string'
+            ? new Threshold(threshold)
+            : threshold;
+
+        return new Match(thresholdObj, redact, count_threshold);
+    }
+}
+
+// Guard class
+export class Guard {
+    readonly name: GuardName;
+    readonly matches: Record<string, Match>;
+    readonly threshold: Threshold;
+    readonly count_threshold: number;
+
+    private constructor(
+        name: GuardName,
+        matches: Record<string, Match>,
+        threshold: Threshold,
+        count_threshold: number = 0
+    ) {
+        this.name = name;
+        this.matches = matches;
+        this.threshold = threshold;
+        this.count_threshold = count_threshold;
+        Object.freeze(this);
+    }
+
+    static create(
+        name: string | GuardName,
+        matches: Record<string, Match> | null = null,
+        threshold: string | Threshold = DEFAULT_THRESHOLD,
+        count_threshold: number = 0
+    ): Guard {
+        // Convert string to GuardName if needed
+        let guardName: GuardName;
+        if (typeof name === 'string') {
+            if (!(name in GuardName)) {
+                const validNames = Object.values(GuardName).join(', ');
+                throw new GuardConfigValidationError(
+                    `Invalid guard name: ${name}. Must be one of: ${validNames}`
+                );
+            }
+            const guardNameInst = GuardName.get(name);
+            if (guardNameInst === undefined) {
+                throw new GuardConfigValidationError(`Invalid guard name: ${name}`);
+            }
+            guardName = guardNameInst;
+        } else {
+            guardName = name;
+        }
+
+        const thresholdObj = typeof threshold === 'string'
+            ? new Threshold(threshold)
+            : threshold;
+
+        return new Guard(
+            guardName,
+            matches || {},
+            thresholdObj,
+            count_threshold
+        );
+    }
+}
+
+// GuardConfig class
+export class GuardConfig {
+    private _parsed_guards: Guard[] = [];
```

Comment:
I believe it is frowned upon in the TypeScript community to prefix things members with `_`. Let's make sure we adhere to some most common best practices at least.

Reply:
done. made the changes at all places to typescript way.